python/pyarrow/array.pxi

```diff
@@ -241,6 +246,13 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
 
     if hasattr(obj, '__arrow_array__'):
         return _handle_arrow_array_protocol(obj, type, mask, size)
+    elif hasattr(obj, '__arrow_c_array__'):
+        if type is not None:
+            requested_type = type.__arrow_c_schema__()
+        else:
+            requested_type = None
+        schema_capsule, array_capsule = obj.__arrow_c_array__(requested_type)
+        return Array._import_from_c_capsule(schema_capsule, array_capsule)
```

Comment:
This should still ensure to cast to `type`, when not None, if the resulting array is not of type `type` (since the "requested schema" is only best effort)?

Reply:
This is a good point. Fixed this.